drkafka/src/main/java/com/pinterest/doctorkafka/KafkaCluster.java

```diff
@@ -254,35 +287,129 @@ public BrokerStats getLatestBrokerStats(int brokerId) {
-        }
-        if (success) {
-          result.put(oosBrokerId, leastUsedBroker);
-          // the broker should not be used again for this topic partition.
-          unusableBrokers.add(leastUsedBroker);
-        } else {
-          LOG.error("Failed to allocate resource to replace {}:{}", oosReplica, oosBrokerId);
-          success = false;
-        }
-      }
+      success &= findNextBrokerForOosReplica(
+          brokerQueue,
+          unusableBrokers,
+          oosReplica.replicaBrokers,
+          result,
+          oosBrokerId,
+          topicPartition,
+          inBoundReq,
+          outBoundReq,
+          preferredBroker
+      );
     }
     // push the brokers back to brokerQueue to keep invariant true
     brokerQueue.addAll(unusableBrokers);
     return success ? result : null;
   }
 
+  /**
+   * Similar to getAlternativeBrokers, but locality aware
+   * @param brokerQueueByLocality a map keeping a priority queue of brokers for each locality
+   * @param oosReplica out of sync replicas
+   * @return a BrokerId to KafkaBroker mapping
+   */
+  public Map<Integer, KafkaBroker> getAlternativeBrokersByLocality(
+      Map<String, PriorityQueue<KafkaBroker>> brokerQueueByLocality,
+      OutOfSyncReplica oosReplica) {
+
+    Map<String, List<Integer>> oosBrokerIdsByLocality = new HashMap<>();
+    for ( int oosBrokerId : oosReplica.outOfSyncBrokers) {
+      String brokerLocality = brokers.get(oosBrokerId).getRackId();
+      oosBrokerIdsByLocality
+          .computeIfAbsent(brokerLocality, l -> new ArrayList<>())
+          .add(oosBrokerId);
+    }
+
+    TopicPartition topicPartition = oosReplica.topicPartition;
+    double inBoundReq = ReplicaStatsManager.getMaxBytesIn(zkUrl, topicPartition);
+    double outBoundReq = ReplicaStatsManager.getMaxBytesOut(zkUrl, topicPartition);
+    int preferredBroker = oosReplica.replicaBrokers.get(0);
+
+    boolean success = true;
+    Map<Integer, KafkaBroker> result = new HashMap<>();
+    Set<KafkaBroker> unusableBrokers = new HashSet<>();
+
+
+    // TODO: fallback to brokers in other locality zones if no brokers are suitable in current locality
+    // Affinity
+    for ( Map.Entry<String, List<Integer>> oosBrokerIdsOfLocality : oosBrokerIdsByLocality.entrySet()) {
+      String oosLocality = oosBrokerIdsOfLocality.getKey();
+      List<Integer> oosBrokerIds = oosBrokerIdsOfLocality.getValue();
+      PriorityQueue<KafkaBroker> localityBrokerQueue = brokerQueueByLocality.get(oosLocality);
+      for( Integer oosBrokerId : oosBrokerIds){
+        success &= findNextBrokerForOosReplica(localityBrokerQueue, unusableBrokers,
+            oosReplica.replicaBrokers,
+            result,
+            oosBrokerId,
+            topicPartition,
+            inBoundReq,
+            outBoundReq,
+            preferredBroker
+        );
+      }
+    }
+
+    // Second pass to fill in replacements without target brokers in locality
```

Comment:
is this still work-in-progress?

Reply:
This is a working PR, however it only implements trivial logic without handling cases such as failover to other localities if no instances are available in the current locality, and spreading out the reassignments to different localities if there are multiple replicas of the same topic partition on the dead broker that is in the same locality. These issues will be addressed in subsequent PRs.